---
title: "Estimating spliced and unspliced abundances and RNA velocity"
author: "Charlotte Soneson"
output:
  html_document:
    keep_md: true
    toc: true
    theme: sandstone
    highlight: tango
date:   2020-05-28
bibliography: rna-velocity.bib
editor_options: 
  chunk_output_type: console
---

```{css, echo = FALSE}
.pythonchunk {
background-color: #faf2dc !important;
}
.rchunk {
background-color: #f5f5f5 !important;
}
.bashchunk {
background-color: #e3eefa !important
}
```

```{r setup, include=FALSE, class.source = "rchunk"}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, eval = TRUE)
options(width = 80)
```

# Introduction and structure

In this part of the course, we will discuss _RNA velocity_ estimation. 
We will cover the underlying mathematical model and assumptions, how the raw data can be quantified in a suitable way, and how the velocity values are estimated and interpreted.

This section uses both R, Python and bash commands. 
The language is indicated by the background color of the respective code chunks. 
R code is represented by <span style="background-color:#f5f5f5">grey</span> boxes, shell commands by <span style="background-color:#e3eefa">blue</span> boxes, and python code by <span style="background-color:#faf2dc">yellow</span> boxes.
Chunks containing output are always white.

# RNA velocity

The gene expression profile of a cell often gives a good indication of the current state of the cell. 
However, being a single snapshot in time, it does not tell us in which _direction_ the cell is heading, or in what state it will be at some point in the near future, and thus it may not be ideal for studying dynamical aspects of biological systems.
_RNA velocity_ (introduced by @La_Manno2018-velocyto) is one approach to addressing this gap, still using a single snapshot of a population of cells and thus without the need to be able to follow the transcriptome of a single cell over time, which is not possible with current scRNA-seq protocols.
In practice, RNA velocity analyses are often summarized by plots such as those shown below (from the [scVelo tutorial](https://scvelo.readthedocs.io/DynamicalModeling.html)): on the left, a vector field overlaid on a low-dimensional embedding, visualizing the 'flow' encoded by the velocities, and on the right, phase plots illustrating single genes.
We will see in this lecture how to generate such plots from raw droplet scRNA-seq data, and how to interpret the results.

![](figures/scvelo_example_dynmod.png)

The RNA velocity is defined as the rate of change of the mature RNA abundance in a cell, and can be estimated from scRNA-seq data by joint modeling of estimated unspliced (pre-mRNA) and spliced (mature mRNA) abundances.
This exploitation of the underlying molecular dynamics of the process sets it apart from other approaches for trajectory analysis, which typically use the similarity of the estimated gene expression profiles among cells to construct a path through the observed data. 
A thorough review of such trajectory inference methods was recently performed by @Saelens2019-trajectory.

The feasibility of the RNA velocity approach is based on the observation by @La_Manno2018-velocyto that, with several commonly used scRNA-seq library preparation protocols, not only exonic, but also intronic and exon/intron boundary-spanning reads are observed, and the insight that considering these in combination with the exonic reads would allow for inference of developmental relationships among cells. 
Similar observations, coupled with a simple differential equation model of transcriptional dynamics, have previously been used for investigation of pre-mRNA dynamics and transcriptional and post-transcriptional regulation of gene expression in exon arrays [@Zeisel2011-premrna] and bulk RNA-seq [@Gaidatzis2015-eisa], as well as estimation of transcription, processing and degradation rates in bulk RNA-seq [@Gray2014-snapshotseq].

## Mathematical model - formulation

Let 

* $u(t)$ denote the abundance of unspliced pre-mRNA
* $s(t)$ denote the abundance of mature mRNA

for a given gene in a given cell at time $t$. 
The system of differential equations underlying the RNA velocity calculations, as formulated by @Zeisel2011-premrna and used in _scVelo_ [@Bergen2019-scvelo], is as follows: 
$$\frac{du(t)}{dt}=\alpha_k(t)-\beta u(t)\\\frac{ds(t)}{dt}=\beta u(t)-\gamma s(t)$$

Let's first break this down a bit: the first equation states that the amount of unspliced RNA increases with a rate $\alpha_k(t)$; this is the rate of transcription, which depends on the _state_ that the gene is in in this particular cell at time $t$ (represented by the index $k$). 
At the same time, due to splicing the unspliced RNA abundance decreases with a rate that is proportional to the current abundance, with a proportionality constant $\beta$, the splicing rate. 

The second equation similarly states that the spliced RNA abundance increases via splicing, with the same amount as the unspliced abundance decreases, and further decreases via degradation, with a proportionality constant $\gamma$. 
The **RNA velocity** of this gene at a given time point $t$, in a given cell, is defined as the value of $ds(t)/dt$, that is, the instantaneous rate of change of the mature mRNA abundance at time $t$. 

## Mathematical model - solution

We can obtain an analytical solution to this set of equations, that is, we can derive an expression for the value of $u(t)$ and $s(t)$ at each point in time (see e.g., @Bergen2019-scvelo). 
To get a better understanding for the model, we will inspect this solution for a few different values of $\alpha_k(t)$, $\beta$ and $\gamma$.
Note that this system of equations is solved separately for _each gene_, so the solutions visualized below represent the abundances for a single gene over time. 

### Constant transcription

First, we consider the case where $\alpha$ is constant over the considered time span.
Also $\beta$ and $\gamma$ are considered to be constant rates. 
We also assume that $u_0=u(0)=0$ and $s_0=s(0)=0$, so that at time 0, this gene is not expressed.
The figures below show three different aspects of the model:

* The lower left panel illustrates the value of $\alpha$ as a function of time. 
* The upper left panel shows the value of $u(t)$ and $s(t)$ as functions of time. 
* The right panel shows a phase plot, with $s(t)$ and $u(t)$ plotted against each other. Each dot corresponds to one time point, and the points have been slightly jittered for better visibility.

```{r plotvelo-alpha1, echo = FALSE, eval = TRUE, fig.width = 7, fig.height = 3}
source("plotVelocityApp/plot_abundances.R")
tf <- data.frame(t = seq(0, 20, by = 0.01)) %>%
  dplyr::mutate(k = ifelse(t == 0, 0, 1)) %>%
  dplyr::mutate(alpha = ifelse(k == 1, 20, 0)) %>%
  dplyr::mutate(beta = 1, 
                gamma = 0.75) %>%
  dplyr::mutate(u = 0, s = 0)
plot_abundances(tf)
```

The first thing we notice is that as the gene starts being transcribed (small values of $t$), both the unspliced and the spliced abundances start increasing.
As expected, the unspliced abundance increases first, followed a bit later by the spliced abundance (since the latter has to go via the unspliced state).

Next, we note that as time passes, both $u(t)$ and $s(t)$ reach a steady state, or equilibrium. 
This indicates that for both the unspliced and spliced pool of molecules, the "influx" is equal to the "outflux", and as a consequence, both $du(t)/dt$ and $ds(t)/dt$ are 0. 
In other words, in this stage, the RNA velocity ($ds(t)/dt$) is zero. 
The exact levels of $u(t)$ and $s(t)$ in the equilibrium phase depend on the values of the parameters $\alpha$, $\beta$ and $\gamma$, but the overall pattern is the same.

Let's see what this looks like in the phase plot on the right hand side. 
The equilibrium point is represented in the top right corner of this plot, and the times leading up to this equilibrium are represented by the points above the diagonal line. 
We can derive the slope of the diagonal line by considering the system of equations above. 
As we saw, in the equilibrium, $$\frac{ds(t)}{dt}=0,$$and thus $$\beta u(t)=\gamma s(t).$$ 
Hence, $$u(t)=(\gamma/\beta) s(t),$$and the slope of the line (assuming that it passes through (0,0), which represents the initial condition) is $\gamma/\beta$.

How can we use this observation in order to estimate RNA velocities when we don't know the values of $\gamma$ and $\beta$ in advance, and need to estimate them from the data? 
Let's assume that the system we are studying is such that we will eventually reach the steady state (we'll see below an example of when this assumption may fail).
Of course, in practice we don't have a time course of abundances of this gene for a single cell, but we have multiple cells, and at our particular snapshot in time, these cells will likely be at different points along the trajectory (corresponding to different values of the underlying time $t$).
Thus, if we assume that the model parameters are the same for all cells (shared splicing and degradation rates), we will still get a phase plot similar to the one above, but each point would correspond to a distinct cell. 
If we then restrict our attention to the cells in the top right corner and bottom left corner of the plot (those assumed to be in steady state), we can estimate the value of $\gamma/\beta$ as the slope of a straight line fitted to these "extreme quantile" data points. 

This is, in essence, the idea behind the approach taken by @La_Manno2018-velocyto and implemented in the _velocyto_ software, and also in the 'steady-state' model in _scVelo_. 
If we fix one of the parameter values (e.g., setting $\beta=1$ as in @La_Manno2018-velocyto, corresponding to an assumption of a shared splicing rate between genes) we can estimate the other one ($\gamma$), and consequently obtain an estimate of the RNA velocity $v$, since $$v=\frac{ds(t)}{dt}=\beta u(t)-\gamma s(t).$$
Notably, these velocities can be derived directly from the phase plot: 

<img src="figures/steady_state_velocity.001.png" width = "50%">

Consider any point along the trajectory. 
By construction, the y coordinate of this point is equal to $u(t)$. 
At the same time, the y coordinate of the diagonal line for the corresponding value of $s(t)$ is $(\gamma/\beta)s(t)$, and thus the difference between the two (the vertical displacement of the point on the trajectory from the diagonal line defined by the steady state) is $$u(t)-\frac{\gamma}{\beta}s(t),$$which is proportional to the velocity as defined above (and, in fact, equal if $\beta=1$).

Once velocities have been estimated, approximate abundances at a future time point $t+\Delta t$ can be estimated e.g. by assuming that the velocity stays constant for this time period: $$s(t+\Delta t) = s(t) + v\Delta t.$$ 

### Transient transcription

Next, let's consider another example, where instead of a constant transcription we consider a transient pattern, where $\alpha>0$ for some time before going back to zero. 

```{r plotvelo-alpha2, echo = FALSE, eval = TRUE, fig.width = 7, fig.height = 3}
source("plotVelocityApp/plot_abundances.R")
tf <- data.frame(t = seq(0, 20, by = 0.01)) %>%
  dplyr::mutate(k = ifelse(t == 0, 0, ifelse(t < 10, 1, 2))) %>%
  dplyr::mutate(alpha = ifelse(k == 1, 20, 0)) %>%
  dplyr::mutate(beta = 1, 
                gamma = 0.75) %>%
  dplyr::mutate(u = 0, s = 0)
plot_abundances(tf)
```

Here, we observe the same type of increase in $u(t)$ and $s(t)$ in the beginning of the time course (of course, the system doesn't know that $\alpha$ will go down to zero in the future). 
When transcription is turned off, both $u(t)$ and $s(t)$ decrease in an exponential fashion. 
Again, the unspliced abundance decreases first, followed by the spliced abundance. 

In the phase plot to the right, we get the same type of pattern as previously for the induction phase (in green). 
Recall that the diagonal line corresponds to the steady-state situation, where $ds(t)/dt = 0$. 
Points above this line represent an induction of gene expression, where the unspliced abundance is higher than the spliced abundance (relative to what we expect under equilibrium). 
Similarly, points below the line (here, in blue), represent a decrease of gene expression, where there is relatively low amounts of unspliced RNA compared to what would be expected under steady state. 

Just as in the case above, if we now assume that each point represents a single cell, we can estimate the ratio $\gamma/\beta$ based on the cells in the top-right and bottom-left parts of the phase plot. 
We can also estimate velocities in the same way as described previously; points above the diagonal line will have positive velocities (induction) and points below the diagonal line will have negative velocities (repression). 

### Short transient transcription

Finally, let's consider a case with the same type of transient transcription as above, but with a much shorter duration. 

```{r plotvelo-alpha3, echo = FALSE, eval = TRUE, fig.width = 7, fig.height = 3}
source("plotVelocityApp/plot_abundances.R")
tf <- data.frame(t = seq(0, 20, by = 0.01)) %>%
  dplyr::mutate(k = ifelse(t == 0, 0, ifelse(t < 3, 1, 2))) %>%
  dplyr::mutate(alpha = ifelse(k == 1, 20, 0)) %>%
  dplyr::mutate(beta = 1, 
                gamma = 0.75) %>%
  dplyr::mutate(u = 0, s = 0)
plot_abundances(tf)
```

Now, there is not enough time to reach the steady state. 
This complicates the estimation of $\gamma/\beta$ based on the phase plot - if we use the points in the upper right part of the plot we will overestimate the ratio. 
In _velocyto_, special approaches are implemented to estimate velocities for genes where all cells are far from steady state.
_scVelo_, instead, implements also a dynamical model, where the full dynamics of the splicing kinetic are solved, thus avoiding the reliance on cells in the steady state to base the estimation of $\gamma/\beta$ on. 
We will discuss the dynamical model in more detail later in this lecture. 


<!-- While the RNA velocity in itself may be hard to interpret, and furthermore we will have one estimate per gene and cell in our data snapshot, a powerful way of using the estimated velocities is to combine them with the current estimates of the processed RNA abundances, and thereby reconstruct the state of the cell at a timepoint in the near future.  -->
<!-- Comparison of these predicted states with the current states of all other cells allows us to infer a developmental order of the cells in our data set, which is typically how RNA velocity is currently used in practice.  -->

## Input data

To solve the system of differential equations, or estimate the velocities, we need to input both spliced (mature mRNA) and unspliced (pre-mRNA) abundances for the genes across all our cells. 
Note that this is not what is typically output from gene expression quantification pipelines (as discussed yesterday), and thus we need a dedicated quantification approach for RNA velocity applications.

### Software tools - overview

Several software packages are able to provide quantifications suitable for RNA velocity analyses. 
Just as for gene expression estimation, the software of choice depends on the scRNA-seq protocol. 
For protocols including UMIs, the software of course needs to be able to handle UMI counting. 
For full-length protocols without UMIs, on the other hand, it is common to use software developed for bulk RNA-seq data. 
Here, we focus on droplet-based protocols with UMIs, such as those provided by 10x Genomics. 
For such data, the most widely used software tools for estimating spliced and unspliced abundances are 

* _velocyto_ [@La_Manno2018-velocyto], the original implementation of the RNA velocity pipeline. Accepts bam files with reads aligned to the genome, e.g., from _CellRanger_ [@Zheng2017-10x].
* _STARsolo_ [@Dobin2013-star], the 'single-cell mode' of the STAR aligner. Operates directly on the raw FASTQ files. 
* _alevin_ [@Srivastava2019-alevin], which is part of _Salmon_ [@Patro2017-salmon], and which you saw already yesterday. Also operates directly on the FASTQ files.
* _kallisto|bustools_, or the _kb-python_ wrapper [@Melsted2019-bustools; @Melsted2019-kb], which leverages _kallisto_ [@Bray2016-kallisto] for quantification of droplet single-cell data. Also operates directly on the FASTQ files. 

Each of these have their own characteristic features, and the quantifications (and subsequent velocities) will depend on the chosen quantification method (see e.g. @Soneson2020-rnavelocity). 

### Software tools - reference sequence

One important characteristic that differ between the software tools just mentioned is the type of reference sequence they operate with. 
_velocyto_ and _STARsolo_ align the reads (or assume that the reads are aligned) to the reference genome, while _alevin_ and _kallisto|bustools_ instead map the reads to an 'expanded' transcriptome. 
The 'expansion' here refers to the inclusion of some form of unspliced or intronic sequences, which are used to represent the pre-mRNA features. 

The genome alignment-based methods (_velocyto_ and _STARsolo_) consider the position where reads are mapped to the genome, and compare those to the positions of annotated exons and introns. 
Reads aligning completely within unambiguously exonic regions are considered 'exonic' (or 'spliced'), and those aligning completely within unambiguously intronic regions are considered 'intronic' (or 'unspliced'). 
Reads overlapping multiple genes are typically discarded. 
All remaining reads, which align in regions that can be either exonic or intronic depending on which isoform one considers, are assigned or discarded as ambiguous according to a (sometimes intricate) set of rules.
The _velocyto_ software, for example, provides several such sets of rules, differing in their 'permissiveness' in allowing ambiguous reads. 

For the transcriptome alignment-based methods (_alevin_ and _kallisto|bustools_), several choices have to be made by the user in order to build the expanded transcriptome to which the reads will be aligned. 
The first decision concerns whether to estimate the pre-mRNA abundances ($u(t)$) as the abundances of the full, unspliced pre-mRNA molecules, or rather (as for the genome alignment-based methods) as the abundance of the _intronic_ parts of a gene. 
This choice has an impact on the resulting quantifications, and the determination of the 'most suitable' approach is still an area of active research and possibly depends on the protocol. 
For 3' tag data such as those from 10x Genomics protocols, initial investigations suggest that there may not be enough information to infer reliably whether fully exonic reads stem from the spliced or the unspliced RNA molecule [@Soneson2020-rnavelocity]. 
For full-length protocols, the situation may be different. 
For this reason, below we will focus on methods defining the pre-mRNA abundance as the abundance of the intronic regions. 

#### Definition of intronic regions

Let's consider the gene in the figure below. 

![](figures/intron_definition_2.001.png)

It has two transcript isoforms, one with two exons and one with three exons. 
The isoforms are partly overlapping. 
While the transcript sequences are straightforward to determine (and also directly provided in the reference transcriptome fasta file), the introns are more ambiguous. 
Essentially we have two choices: either we define the introns separately for each isoform, or we first collapse the isoforms into the union of their exonic regions, and define the introns as the regions that are within the gene locus, but not included in the exon union.

As you can see in the two illustrations above, with the "separate" approach the introns of a gene can overlap with each other and with the exonic regions, while this is not allowed in the "collapse" approach. 
Another way to see this is that with the "separate" approach we are considering exons and introns on an equal footing, whereas with the "collapse" approach reads in ambiguous regions will be considered exonic. 

Since 'intronic' reads can be either fully intronic or partly exonic and partly intronic (overlapping an exon/intron boundary), we add a flanking sequence to each extracted intron to allow for the latter type of reads. 
The length of this flanking sequence can be anywhere up to the length of the RNA read minus 1, depending on how much intronic overlap we want to require in order for a read to be considered potentially intronic. 

In the lower part of the figure above, we summarize the features (transcripts + introns) that will be indexed for each of the two approaches. 
In each case, the 'spliced' abundance of the gene is defined as the sum of the abundances of the individual transcripts, and the 'unspliced' gene abundance is defined as the sum of the abundances of the introns.

### Software tools - other characteristics

There are also differences between the quantification tools listed beyond their choice of reference sequence to align or map the reads to. For example:

* Only _alevin_ will consider reads that could stem from more than one gene in the quantification, while the other tools will discard these reads.
* _kallisto|bustools_ does not take into account the strandedness of the reads (e.g., from 10x Genomics), and thus quantifies the abundances as if the library was unstranded. 
This has several consequences for the quantification. 
First, consider the stituation of a gene overlapping another gene on the opposite strand. 
Reads falling in the overlapping region will be considered ambiguous, and discarded, although the fact that the 10x Genomics libraries are stranded would have allowed them to be unambiguously assigned. 
On the other hand, counting reads from both strands allow the inclusion of intronic reads stemming from discordant 'internal priming' on polyT sites. 
This phenomenon was observed by @La_Manno2018-velocyto, and suggests that in addition to the expected peaks around intronic polyA sites, there are also peaks with reads on the opposite strand, around polyT sites. 
If only reads aligning to the same strand as the introns are counted, all these peaks are discarded. 

In order to better understand some of these differences, we show below a few examples where different methods provide different quantifications (from @Soneson2020-rnavelocity):

* Chkb - overlapping features on the same strand. In this case, only _alevin_ assigns a non-zero UMI count (and _STARsolo-diff_, which defines the intronic count as the difference between a "gene body count" and the regular gene expression). 

![](figures/pancreas_Chkb.png)

* Rassf1 - overlapping features on different strands. 
Whether or not the tool accounts for the strandedness of the reads makes a difference.

![](figures/pancreas_Rassf1.png)

* Tspan3 - many ambiguous regions. 
The way that the introns are defined makes a substantial difference. 
The intronic count is much higher with the 'separate' intron definition approach.

![](figures/pancreas_Tspan3.png)

These differences between counts obtained by different methods propagate also to the estimated velocities, and can affect the biological interpretation of the final results. 

# Practical workflow

In this tutorial, we show how the 'spliced' and 'unspliced' count matrices can be generated using [_alevin_](https://salmon.readthedocs.io/en/latest/alevin.html) [@Srivastava2019-alevin] and [_tximeta_](https://bioconductor.org/packages/tximeta/) [@Love2020-tximeta].
We then illustrate how the processed data can be read into python and used as input for RNA velocity estimation with [_scVelo_](https://scvelo.readthedocs.io/) [@Bergen2019-scvelo].
It is worth keeping in mind that this is _not_ the only way that these steps can be performed, that this is a very active research area, and that it's plausible that some of the recommendations below will need to be updated as we learn more about this field and as software packages are developed further. 
However, the conceptual overview of the different steps and the guidance on the interpretation of the output is still applicable. 

Recall that the code blocks represent both R, python and bash commands, and the language is indicated by the background color of the respective code chunks. 
R code is represented by <span style="background-color:#f5f5f5">grey</span> boxes, shell commands by <span style="background-color:#e3eefa">blue</span> boxes, and python code by <span style="background-color:#faf2dc">yellow</span> boxes.
Chunks containing output are always white.

## Preparation

We start by loading the R packages that we will need for the analyses that follow. 
All packages are available via [CRAN](https://cran.r-project.org/) or [Bioconductor](https://bioconductor.org/) (>=3.11), and can be installed via 

```{r installpkg, eval = FALSE, class.source = "rchunk"}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("<packagename>")
```

```{r loadpkg, class.source = "rchunk"}
suppressPackageStartupMessages({
    library(Biostrings)
    library(BSgenome)
    library(eisaR)
    library(GenomicFeatures)
    library(SummarizedExperiment)
    library(SingleCellExperiment)
    library(tximeta)
    library(rjson)
    library(reticulate)
    library(scater)
    library(scran)
    library(fishpond)
})
```

We will practice generating the [_Salmon_](https://salmon.readthedocs.io/en/latest/salmon.html) index required for _alevin_ quantification using the same small reference files that you used yesterday to create the index for the regular gene expression quantification. 
Here, we first set the path to the data (`datadir`), as well as to the folder where we will store the generated index and quantifications (`outdir`).

```{r setpaths, class.source = "rchunk", eval = TRUE}
if (file.exists("/home/rstudio/adv_scrnaseq_2020")) {
  datadir <- "/home/rstudio/adv_scrnaseq_2020/data/spermatogenesis_subset"
  outdir <- "/home/rstudio/adv_scrnaseq_2020/data/spermatogenesis_subset/txintron"
} else {
  datadir <- "data/spermatogenesis_subset"
  outdir <- "data/spermatogenesis_subset/txintron"
}
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)
Sys.setenv(datadir = datadir, outdir = outdir)
```

```{bash listfiles, class.source = "bashchunk"}
## If run in a console
## datadir=/home/rstudio/adv_scrnaseq_2020/data/spermatogenesis_subset
## outdir=/home/rstudio/adv_scrnaseq_2020/data/spermatogenesis_subset/txintron
## Check what is included in the data directory
ls $datadir
```

## Velocity quantification with _alevin_

### Step 1. Generate reference fasta files

In order to quantify both exonic and intronic abundances with _alevin_, we need to provide a reference fasta file with both types of sequences, as discussed previously. 
Several R/Bioconductor packages implement the extraction of transcript and intron sequences from the genome sequence. 
Here, we will use the [_eisaR_](https://bioconductor.org/packages/eisaR/) package, but equivalent functionality is available in the [_BUSpaRse_](https://bioconductor.org/packages/BUSpaRse/) package, and can be replicated e.g. using basic functions from the [_GenomicFeatures_](https://bioconductor.org/packages/GenomicFeatures/) and [_BSgenome_](https://bioconductor.org/packages/BSgenome/) packages. 

The first step in the feature extraction with _eisaR_ is to extract a `GRangesList` object containing the genomic coordinates of each annotated transcript and intron. 
In this example, we use the "separate" approach to define introns separately for each transcript, and add a flank length of 90nt to each intron, since the RNA read length of this data set is 91nt.

```{r get-featureranges, class.source = "rchunk"}
grl <- eisaR::getFeatureRanges(
  gtf = file.path(datadir, "GRCm38.gencode.vM21.chr18.chr19.gtf"),
  featureType = c("spliced", "intron"), 
  intronType = "separate", 
  flankLength = 90L, 
  joinOverlappingIntrons = FALSE, 
  verbose = TRUE
)
```

The resulting object is a `GRangesList`, each element of which is a feature (transcript or intron) to be indexed.
Transcripts can be either mono-exonic or multi-exonic; for the latter, the corresponding entry in the `GRangesList` has multiple ranges.
Introns are always "mono-exonic". 
The feature ID of an intron will be set to `<transcript_id>-I<intron_number>` (or just `<transcript_id>-I` for the first intron in the transcript) with the 'separate' approach. 
With the 'collapse' approach, the `-I<intron_number>` suffix will instead be prefixed by the gene ID. 
For each feature, in addition to the range information _eisaR_ returns a gene ID. 
For introns, this will be the original gene ID, with a suffix `-I`. 
This is required for _alevin_ later, since it will aggregate counts on the gene level, and we want to get a separate count for the exonic and intronic features of each gene. 

```{r inspect-featureranges, class.source = "rchunk"}
head(grl, 2)
tail(grl, 2)
```

Let's see how many transcripts and introns there are among the features (recall that introns have a `-I` in their name)

```{r feature-types, class.source = "rchunk"}
table(grepl("-I", names(grl)))
```

This shows that most of the features are in fact introns, as we might expect, since a single transcript can contribute multiple introns. 
As a further sanity check we can control that there are the same number of transcript features as in the transcriptome fasta file from GENCODE (also subset to chromosomes 18 and 19):

```{r check-txome, class.source = "rchunk"}
txome <- Biostrings::readDNAStringSet(
  file.path(datadir, "GRCm38.gencode.vM21.chr18.chr19.txome.fa")
)
length(txome)
```

The `grl` object contains the genomic ranges for the features, but _Salmon_ needs a fasta file with sequences for the indexing. 
Thus, given the positional information and the genome sequence, we extract these sequences and write them to a fasta file for later indexing with _Salmon_.


```{r extract-seqs, class.source = "rchunk"}
genome <- Biostrings::readDNAStringSet(
  file.path(datadir, "GRCm38.gencode.vM21.chr18.chr19.genome.fa")
)
names(genome) <- sapply(strsplit(names(genome), " "), .subset, 1)
genome
seqs <- GenomicFeatures::extractTranscriptSeqs(
  x = genome, 
  transcripts = grl
)
seqs
Biostrings::writeXStringSet(
    seqs, 
    filepath = file.path(
      outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.fa"
    )
)
```

We can also make sure that the transcripts extracted by _eisaR_ are in fact identical to those downloaded from GENCODE (albeit in a different order).

```{r check-txs, class.source = "rchunk"}
names(txome) <- sapply(strsplit(names(txome), "\\|"), .subset, 1)
all(names(txome) %in% names(seqs))
all(seqs[names(txome)] == txome)
```

While the fasta file we just generated is all that _Salmon_ requires for the indexing, we will generate a few more related reference files, to make downstream analysis easier. 
First, in order to enable reading the estimated abundances with _tximeta_, while automatically recognizing the underlying transcriptome, we export the expanded annotation to a GTF file.
This will later be used to create a linked transcriptome for _tximeta_.

```{r export-gtf, class.source = "rchunk"}
eisaR::exportToGtf(
  grl, 
  filepath = file.path(
    outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.gtf"
  )
)
```

Since _alevin_ quantifies spliced and unspliced features jointly, we will eventually also need to split the imported abundances by feature type, to generate two count matrices (with spliced and unspliced counts) rather than one joint matrix. 
The splitting needs to be done in such a way that we can still match up a spliced feature with the corresponding unspliced feature, that is, that the rows of the two resulting matrices are the same. 
To help with this, the metadata of the `GRangesList` object contains a data frame with corresponding spliced and unspliced gene IDs, which we write to a text file for later use.

```{r check-corrgene, class.source = "rchunk"}
head(metadata(grl)$corrgene)
```

```{r write-features, class.source = "rchunk"}
write.table(
    metadata(grl)$corrgene, 
    file = file.path(
      outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.features.tsv"
    ),
    row.names = FALSE, col.names = TRUE, quote = FALSE, sep = "\t"
)
```

Finally, we create a text file mapping transcript and intron identifiers to the corresponding gene identifiers, which _alevin_ will use to generate the gene-level count matrix.

```{r write-tx2gene, class.source = "rchunk"}
df <- eisaR::getTx2Gene(
    grl, 
    filepath = file.path(
      outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.tx2gene.tsv"
    )
)
head(df)
tail(df)
```

Now, the folder `r outdir` contains all the reference files we need for further indexing with _Salmon_, quantification with _alevin_ and import into R with _tximeta_. 

```{r list-txintron, class.source = "rchunk"}
dir(outdir)
```


### Step 2. Index the reference features

After creating the fasta file with transcript and intron sequences as above, we index it using _Salmon_. 
We add the full genome as decoy sequences (see @Srivastava2019-mm for more details). 
This is recommended in order to avoid reads truly originating from intergenic regions being assigned to a suboptimal transcriptome location. 
However, the effect of including decoys is typically smaller when both transcripts and introns are being quantified than for 'regular' gene expression quantification, since in the former case a larger fraction of the genome is already covered by the features of interest.

```{r set-salmonpath, echo = FALSE}
if (file.exists("/Users/charlotte/miniconda3/envs/salmon/bin/salmon")) {
  Sys.setenv(PATH = paste(Sys.getenv("PATH"), 
                          "/Users/charlotte/miniconda3/envs/salmon/bin", 
                          sep = ":"))
}
```

```{bash index-genome, class.source = "bashchunk", results = "hide"}
grep ">" $datadir/GRCm38.gencode.vM21.chr18.chr19.genome.fa | \
cut -d ">" -f 2 | cut -d " " -f 1 \
> $datadir/GRCm38.gencode.vM21.chr18.chr19.genome.chrnames.txt

salmon index \
-t <(cat $outdir/GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.fa $datadir/GRCm38.gencode.vM21.chr18.chr19.genome.fa) \
-i $outdir/GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.sidx \
--gencode -p 2 \
-d $datadir/GRCm38.gencode.vM21.chr18.chr19.genome.chrnames.txt
```

```{bash list-index, class.source = "bashchunk"}
ls $outdir
```

At this stage, we will also create a _linked transcriptome_ with _tximeta_. 
When importing 'regular' gene expression abundances with _tximeta_, you noticed that it automatically recognized a large collection of references. 
However, since the reference sequences used here (transcripts and introns) do not directly correspond to a fasta file downloaded from any official source, we have to generate a linked transcriptome ourselves. 
This allows _tximeta_ to recognize the reference annotation when reading the _alevin_ quantification, and automatically annotate the resulting `SummarizedExperiment` object.

```{r make-linkedtxome, class.source = "rchunk"}
tximeta::makeLinkedTxome(
  indexDir = file.path(
    outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.sidx"
  ), 
  source = "GENCODEtxintron", genome = "GRCm38", 
  organism = "Mus musculus", release = "M21", 
  fasta = file.path(
    outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.fa"
  ), 
  gtf = file.path(
    outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.gtf"
  ), 
  write = TRUE, 
  jsonFile = file.path(
    outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.json"
  )
)
```

Let's see what is stored in the json file:

```{r explore-json, class.source = "rchunk"}
rjson::fromJSON(file = file.path(
  outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.json"
))
```

### Step 3. Quantify with alevin

After generating the index, we quantify exonic and intronic abundances with _alevin_. 
This is done exactly like for regular gene expression estimation. 
Recall that in the transcript-to-gene text file generated above, transcripts and introns from the same original gene will be artificially assigned to different genes (the intron gene will have a `-I` suffix). 
This allows _alevin_ to aggregate exonic and intronic reads separately.

```{bash quantify, class.source = "bashchunk", results = "hide"}
salmon alevin -l ISR \
-i $outdir/GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.sidx \
-1 $datadir/AdultMouseRep3sub1M_S1_L001_R1_001.fastq.gz \
-2 $datadir/AdultMouseRep3sub1M_S1_L001_R2_001.fastq.gz \
-o $outdir/alevin_GRCm38.gencode.vM21.chr18.chr19.transcripts.introns \
-p 2 --chromium --dumpFeatures \
--tgMap $outdir/GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.tx2gene.tsv
```

```{bash ls-alevin, class.source = "bashchunk"}
ls -R $outdir/alevin_GRCm38.gencode.vM21.chr18.chr19.transcripts.introns
```

### Step 4. Import abundances into R with _tximeta_

The _tximeta_ package can be used to import the _alevin_ quantifications into R, and generate a `SummarizedExperiment` object. 
We first load the linked transcriptome generated above, and then read the _alevin_ output. 

```{r read-tximeta, class.source = "rchunk"}
tximeta::loadLinkedTxome(file.path(
  outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.json"
))
txi <- tximeta::tximeta(coldata = data.frame(
  names = "AdultMouseRep3",
  files = file.path(
    outdir, "alevin_GRCm38.gencode.vM21.chr18.chr19.transcripts.introns", 
    "alevin/quants_mat.gz"
  ), 
  stringsAsFactors = FALSE
), type = "alevin")
```

The `txi` object contains a single assay ('counts') containing both spliced and unspliced abundances. 
In order to calculate RNA velocities, we need to split this into two matrices, one with spliced and one with unspliced abundances, with corresponding rows. 
This can be done using the `splitSE()` function from _tximeta_, providing the data frame linking spliced and unspliced gene identifiers that we created above. 
Note that _tximeta_ version 1.5.30 or later is required for this step.

```{r splitse, class.source = "rchunk"}
cg <- read.delim(file.path(
  outdir, "GRCm38.gencode.vM21.chr18.chr19.transcripts.introns.features.tsv"
), header = TRUE, as.is = TRUE)

## Rename the 'intron' column 'unspliced' to make assay names compatible with scVelo
colnames(cg)[colnames(cg) == "intron"] <- "unspliced"

## Split the SummarizedExperiment object and convert to SingleCellExperiment
txis <- tximeta::splitSE(txi, cg, assayName = "counts")
txis <- as(txis, "SingleCellExperiment")

## Remove metadata to allow anndata2ri to export AnnData object later
metadata(txis) <- list()
```

At this point, the `txis` object contains all the information required for RNA velocity analysis. 

```{r printtxis, class.source = "rchunk"}
txis
```

## RNA velocity analysis with _scVelo_

Once we have the spliced and unspliced abundance estimates, we can run, e.g., [_scVelo_](https://scvelo.readthedocs.io/) [@Bergen2019-scvelo] to estimate and visualize RNA velocities. 
The code in this section will be run in python.
First, we import the modules that we will use in the analysis and set the path to use for RNA velocity output.
In the course environment, all python modules are available from the system python installation. 
However, on your own system, you may need to activate a suitable environment (or link [_reticulate_](https://rstudio.github.io/reticulate/) to it if running via R).

```{r use-virtualenv, class.source = "rchunk", echo = FALSE}
if (file.exists("/Users/charlotte/miniconda3/envs/scvelo")) {
  reticulate::use_condaenv("/Users/charlotte/miniconda3/envs/scvelo")
}
```

```{python load-modules, class.source = "pythonchunk"}
import scanpy as sc
import sys
import numpy as np
import anndata
import scvelo as scv
import matplotlib
import pandas as pd
from pathlib import Path
from os import path
```

```{python set-velodir, class.source = "pythonchunk"}
## Path to data to use for RNA velocity calculations
if (path.exists("/home/rstudio/adv_scrnaseq_2020")):
  velodir = Path('/home/rstudio/adv_scrnaseq_2020/data/spermatogenesis_rnavelocity')
else:
  velodir = Path('data/spermatogenesis_rnavelocity')
```

We also set a few general parameters, prettifying plots and defining the path to the directory where _scVelo_ will save plots if requested. 

```{python set-params, class.source = "pythonchunk"}
## Set matplotlib renderer (using the Anti-Grain Geometry engine)
matplotlib.use('AGG')
#scv.settings.set_figure_params('scvelo')

scv.settings.figdir = velodir
scv.settings.plot_prefix = "scvelo_"
```

### Data import

So far, we have used a small example data set to illustrate how to generate the reference index and quantify spliced and unspliced abundances with _alevin_. 
Here, we will use the full data set, which was processed as illustrated above, but using the full FASTQ files and the complete GENCODE reference files. 
This data set is provided as a `SingleCellExperiment` object.
In addition to the steps outlined above, we have added a `counts` assay containing the counts obtained from a 'regular' gene expression quantification with _alevin_.
We have also applied a standard scRNA-seq workflow to normalize and log-transform these counts, and applied PCA, tSNE and UMAP.
The purpose of these steps is to provide a way to eventually visualize the velocities on top of the same reduced dimension representations that we would use for the other parts of our scRNA-seq workflow. 

#### Conversion to AnnData

In order to work with our data in python, we need to convert the object from a `SingleCellExperiment` to an [`AnnData`](https://anndata.readthedocs.io/en/stable/anndata.AnnData.html) object, which is the corresponding python container. 
This is not directly handled by _reticulate_, since these are not standard R/python object classes. 
A few R packages currently provide conversion functionality; however, [_Seurat_](https://satijalab.org/seurat/) currently does not export `AnnData` objects, and [_sceasy_](https://github.com/cellgeni/sceasy) requires an old version of the _anndata_ python module. 
Thus, here we will illustrate how the conversion can be done using the [_anndata2ri_](https://github.com/theislab/anndata2ri) python module, which uses [_rpy2_](https://pypi.org/project/rpy2/) to communicate between python and R, and implements a conversion function. 
Depending on your setup, unfortunately, it may be problematic to call this python conversion function from within R via _reticulate_ (since there is already an R session running, and _rpy2_ will try to create another one). 
Below, we therefore provide the code that you would run to convert the `SingleCellExperiment` object to an `AnnData` object (this would work if run directly in a python session, but may be problematic if run from within R via _reticulate_), but we also provide the already-converted object, stored in an `.h5ad` file.

```{r python load-data, class.source = "pythonchunk", eval = FALSE}
## DO NOT RUN THIS FROM THE COURSE RSTUDIO SESSION
#import anndata
#import anndata2ri
#from rpy2.robjects import r
#anndata2ri.activate()
#adata = r(f'sce <- readRDS("data/spermatogenesis_rnavelocity/AdultMouseRep3_alevin_GRCm38.gencode.vM21.spliced.intron.fl90.gentrome.k31_sce_nometa.rds")')
#adata.write("data/spermatogenesis_rnavelocity/AdultMouseRep3_alevin_GRCm38.gencode.vM21.spliced.intron.fl90.gentrome.k31_sce_nometa.h5ad")
```

```{python load-h5ad, class.source = "pythonchunk"}
adata = anndata.read(
velodir/'AdultMouseRep3_alevin_GRCm38.gencode.vM21.spliced.intron.fl90.gentrome.k31_sce_nometa.h5ad'
)
```

#### The AnnData object

The _AnnData_ object contains a data matrix (`X`, which will be defined by the first assay in the `SingleCellExperiment` object), any other assays contained in the `SingleCellExperiment` (denoted `layers`; here, spliced, unspliced and logcounts). 
It also contains cell-level annotations (`adata.obs`), variable annotations (`adata.var`, empty in this data set), reduced dimension representations (`adata.obsm`) and unstructured annotations (`adata.uns`). 
The names of observations and variables can be accessed via `adata.obs_names` and `adata.var_names`, respectively. 

```{python check-anndata, class.source = "pythonchunk"}
adata
adata.layers
adata.obs
adata.uns
adata.obs_names
adata.var_names
```

We can plot or print the proportion of counts in the data set that are assigned to the unspliced targets:

```{python proportions, class.source = "pythonchunk"}
## Plot proportion of unspliced counts
scv.utils.show_proportions(adata)
scv.pl.proportions(adata, groupby = "celltype")
```


### Gene filtering

The imported `AnnData` object contains all the quantified genes. 
We use _scVelo_ to select highly variable genes to use as the basis for the velocity calculations. 
First, we exclude all genes for which the total (spliced and unspliced) count, added up across cells with non-zero spliced and unspliced count, don't exclude a given threshold (here, 30).
Next, we normalize each layer separately. 
The `enforce = True` is required since _scVelo_ contains logic to determine whether the provided data is already normalized, by checking whether the provided values are integers. 
Since the counts from _alevin_ are not necessarily integers, the data may be incorrectly identified as pre-normalized.
After the normalization, highly variable genes are selected (based on the normalized `X` matrix), by default (the 'seurat' flavor) by binning the genes into bins of similar expression and selecting genes with the highest variance within each bin.
Finally, the `X` matrix is log-transformed. 

It is worth noting that the set of genes used to estimate RNA velocities can have a strong effect on the final interpretation. 
Especially in data sets containing multiple, competing dynamical signals, focusing only on genes involved in one of them can give a very different overall picture than including all the highly variable genes. 

```{python filter, class.source = "pythonchunk"}
scv.pp.filter_genes(adata, min_shared_counts = 30)
scv.pp.normalize_per_cell(adata, enforce = True)
scv.pp.filter_genes_dispersion(adata, n_top_genes = 2000)
scv.pp.log1p(adata)
adata
```

### Velocity estimation

_scVelo_ implements three different models for estimating velocities. 
These are described in much more detail in @Bergen2019-scvelo.

* **Steady-state**. Essentially the same model that was originally proposed by @La_Manno2018-velocyto. 
Cells in steady state are inferred from the phase plot, and the value of $\gamma$ is calculated as the slope of the straight line fitted to these cells.
This model assumes that all genes share a common splicing rate, and that the induction and repression periods are long enough to reach steady state.
* **Stochastic**. An extension of the steady-state model, treating transcription, splicing and degradation as probabilistic events.
This model was shown to improve the performance of the steady-state model, without significantly increasing the computational time.
* **Dynamical**. The most complex (and most time-consuming) model, with the best performance. 
In this likelihood-based framework, the reaction rate parameters ($\alpha_k$, $\beta$, $\gamma$) as well as latent variables (time and transcriptional state) are estimated for each gene using an EM algorithm. 
The model assumes the existence of four different transcriptional states - two steady-state (high and low), induction and repression. 
The EM algorithm iterates between estimating the latent time of a cell (the 'position' of the cell along the phase space trajectory) and assigning it a transcriptional state, and optimizing the values of the parameters (see Figure below from @Bergen2019-scvelo). 
The likelihood is obtained by assuming that the observations follow a normal distribution:$$x_i^{obs}\sim N((\hat{u}(t), \hat{s}(t)), \sigma^2).$$

![](figures/BergenFig1.jpg)

Here, we will focus on the dynamical model, since it is generally the most accurate, and although it's a bit slower than the other methods, usually it's not prohibitively slow.

The first step is to estimate moments (mean and uncentered variance) of the abundances for each cell. 
The moments are calculated across a small set of neighboring cells (in the PCA space), for increased stability. 
Two layers Ms and Mu, which are the first order moments (means) for spliced and unspliced abundances, are added to `adata`.
The neighbors are stored as well

```{python get-moments, class.source = "pythonchunk"}
scv.pp.moments(adata, n_pcs = 30, n_neighbors = 30)
adata
adata.uns['neighbors']
```

Next, we estimate the velocities. 
First, the `tl.recover_dynamics` function fits the model and infers transcription rates, splicing rates and degradation rates for each gene, as well as cell-specific latent times and transcriptional states.
This step is only required for the dynamical model.

```{python recover-dynamics, class.source = "pythonchunk"}
scv.tl.recover_dynamics(adata)
adata
```

This step adds several columns to `adata.var` (see [https://scvelo.readthedocs.io/DynamicalModeling.html](https://scvelo.readthedocs.io/DynamicalModeling.html)), including:

* $R^2$ of the linear fit to the steady-state cells in the phase plot (`fit_r2`). Note that this can be negative, if the obtained fit is worse than just using a straight line at the mean. This is used to determine which genes are used for the downstream analysis and projection of velocities into a low-dimensional space.
* transcription rate estimates (`fit_alpha`)
* splicing rate estimates (`fit_beta`)
* degradation rate estimates (`fit_gamma`)
* estimates of switching time points (`fit_t_`)
* the likelihood value of the fit (`fit_likelihood`), averaged across all cells. The likelihood value for a gene and a cell indicates how well the cell is described by the learned phase trajectory.

Once the kinetic rate parameters are estimated, the `tl.velocity` function estimates the actual velocities based on these. 
This adds a `velocity` layer to the `adata` object, and the `velocity_genes` column in `adata.var`.
This column indicates whether the fit for a gene is considered 'good enough' for downstream use. 
Specifically, it requires `fit_r2` and `fit_gamma` to exceed certain thresholds, and that both the unspliced and spliced mean values are nonzero. 

```{python run-scvelo, class.source = "pythonchunk"}
scv.tl.velocity(adata, mode = 'dynamical', min_r2 = 0.01, min_likelihood = 0.001)
adata
adata.var['velocity_genes'].sum()
```

We can plot the distributions of the estimated kinetic rates, for the retained 'velocity genes'.

```{python plot-rate-distrs, class.source = "pythonchunk"}
df = adata.var
df = df[df['velocity_genes'] == True]

kwargs = dict(xscale = 'log', fontsize = 30)
with scv.GridSpec(ncols = 5) as pl:
    pl.hist(df['fit_alpha'], xlabel = 'transcription rate', **kwargs)
    pl.hist(df['fit_beta'] * df['fit_scaling'], xlabel = 'splicing rate', **kwargs)
    pl.hist(df['fit_gamma'], xlabel = 'degradation rate', **kwargs)
    pl.hist(df['fit_r2'], xlabel = 'fit r2', fontsize = 30)
    pl.hist(df['fit_likelihood'], xlabel = 'fit likelihood', fontsize = 30)

scv.get_df(adata, 'fit*', dropna = True).head()
```

### Generating the velocity graph

At this point, we have estimated the velocities - these are vectors in a $K$-dimensional space, where $K$ is the number of retained genes. 
In order to use these velocities for downstream applications, such as estimating the future state of an individual cell or generating low-dimensional visualizations, we next estimate a so called _velocity graph_. 
To this end, _scVelo_ calculates cosine similarities between the velocity vector for each cell and the displacement vector from that cell to each other (neighboring) cell:
$$\pi_{ij}=cos\angle(\delta_{ij}, v_i),$$where $\delta_{ij}=s_j-s_i$ is the displacement vector from cell $i$ to cell $j$ in gene expression space, and $v_i$ is the velocity vector of cell $i$.
The cosine similarity takes values between -1 and +1, and a large value indicates that cell $i$ has a high probability of transitioning towards cell $j$.
The velocity graph is stored in the `uns` slot of the `adata` object, and is represented by a sparse $N\times N$ matrix (where $N$ is the number of cells).
There is also a `velocity_graph_neg`, which is a matrix of the same size as `velocity_graph`, containing the negative cosine similarities

```{python velo-graph, class.source = "pythonchunk"}
## Calculate the velocity graph
scv.tl.velocity_graph(adata)

## Note that adata.uns now contains the velocity graph
adata
adata.uns['velocity_graph']
adata.uns['velocity_graph'].todense()
```

The velocity graph can be plotted with the dedicated `pl.velocity_graph` function, and overlaid on top of any existing reduced dimension representation in the object (here, the tSNE).

```{python plot-velo-graph, class.source = "pythonchunk"}
## Plot the velocity graph
scv.pl.velocity_graph(adata, threshold = 0.1, basis = 'tsne', 
                      color = 'celltype', title = '')
```

The graph can also be used to estimate the most likely cell transitions, and thereby "trace" the likely path of a given starting cell. 

```{python trace-descendants, class.source = "pythonchunk"}
x, y = scv.utils.get_cell_transitions(adata, basis = 'tsne', starting_cell = 70)
ax = scv.pl.velocity_graph(adata, basis = 'tsne', c = 'lightgrey', edge_width = 0.05, show = False)
ax = scv.pl.scatter(adata, x = x, y = y, s = 120, c = 'ascending', cmap = 'gnuplot', ax = ax)
```

### Visualizing the velocities in low dimension

In order to visualize the velocities in a lower-dimensional embedding, we convert the cosine similarities in the velocity graph to _transition probabilities_ of cell-to-cell transitions by applying an exponential kernel:
$$\tilde{\pi}_{ij}=\frac{1}{z_i}exp(\frac{\pi_{ij}}{\sigma_i^2}).$$
The $z_i$ are normalization factors and $\sigma_i$ is an adaptive kernel width parameter.

```{python get-trans-matrix, class.source = "pythonchunk"}
scv.utils.get_transition_matrix(adata)
scv.utils.get_transition_matrix(adata).todense()
```

These transition probabilities are used to project the velocities into a low-dimensional embedding. 
This is achieved by weighting the normalized displacement vectors from a cell to all other cells in the low-dimensional space by the transition probabilities for cell, and taking the resulting weighted average as the low-dimensional velocity vector.
More precisely, if $$\tilde{\delta}_{ij}=\frac{\tilde{s}_j-\tilde{s}_i}{\|\tilde{s}_j-\tilde{s}_i\|}$$are the normalized displacement vectors in the low-dimensional embedding, the embedded velocity is estimated by $$\tilde{v}_i=\sum_{j\neq i}\left(\tilde{\pi}_{ij} - \frac{1}{n}\right)\tilde{\delta}_{ij}.$$
These projected velocities can be calculated with `tl.velocity_embedding`, and visualized in different ways:

* on a cell-by-cell level (with `pl.velocity_embedding`)
* as gridlines (with `pl.velocity_embedding_grid`)
* as streamlines (with `pl.velocity_embedding_stream`)

```{python plot-velo-emb, class.source = "pythonchunk"}
scv.pl.velocity_embedding(adata, basis = 'tsne', color = 'celltype', title = '')
scv.pl.velocity_embedding_grid(adata, basis = 'tsne', color = 'celltype', title = '')
scv.pl.velocity_embedding_stream(adata, basis = 'tsne', color = 'celltype', title = '')
```

```{python get-proj-velo, class.source = "pythonchunk"}
scv.tl.velocity_embedding(adata, basis = 'tsne', all_comps = False, autoscale = False)
pd.DataFrame(adata.obsm['velocity_tsne'], index = adata.obs.index)
```

### Pseudotime and latent time

It is often of interest to obtain an ordering of the cells along a trajectory. 
_scVelo_ provides two different approaches for this: pseudotime and latent time. 

* The velocity pseudotime is obtained via a diffusion random walk on the velocity graph, and measures how many steps (on average) it takes to reach a given cell from one of the root cells.
The root cells are obtained from the transition matrix.
* The latent time is obtained from the transcriptional dynamics fit, by relating gene-specific times (position along the phase curve) to a "universal" latent time, shared across genes. 
Usually, the latent time captures the _actual_ time better than pseudotime.

```{python pseudotime, class.source = "pythonchunk"}
scv.tl.velocity_pseudotime(adata)
scv.pl.scatter(adata, basis = "tsne", color = "velocity_pseudotime", color_map = "gnuplot")

scv.tl.latent_time(adata)
scv.pl.scatter(adata, basis = "tsne", color = "latent_time", color_map = "gnuplot")
```

### Single-gene plots

Above we saw how to interpret the velocities by projecting them into a low-dimensional representation. 
Next, we will look at individual genes via their phase portraits. 
This is often useful in order to understand how the velocities are affected/supported by specific genes. 
We can of course plot genes that we are already familiar with and that we know are related to the process of interest. 
We can also extract genes with particularly strong influence on the velocity results.
'Driver genes', which display a strong dynamic behavior, can for example be detected via their high likelihoods in the dynamic model.

```{python top-genes, class.source = "pythonchunk"}
top_genes = adata.var['fit_likelihood'].sort_values(ascending = False).index[:100]
top_genes
```

We can plot the expression of these genes in a heatmap, ordering the cells by the inferred latent time.

```{python heatmap, class.source = "pythonchunk"}
scv.pl.heatmap(adata, var_names = top_genes, sortby = 'latent_time', n_convolve = 50, col_color = 'celltype')
```

We can also show the phase portraits of the individual genes.

```{python phase-portraits, class.source = "pythonchunk"}
scv.pl.scatter(adata, basis = top_genes[:10], legend_loc = 'none', size = 80,
               frameon = True, ncols = 5, fontsize = 20, color = 'celltype')
```

Or plot their expression against the inferred latent time.

```{python expr-time, class.source = "pythonchunk"}
scv.pl.scatter(adata, x = 'latent_time', y = top_genes[:10], ncols = 5, 
               frameon = True, color = 'celltype', fontsize = 20)
```

_scVelo_ also provides nice summary plots, combining phase portraits, velocity maps and expression plots. 

```{python summary-plots, class.source = "pythonchunk"}
scv.pl.velocity(adata, top_genes[:1], ncols = 1, basis = 'tsne', color = 'celltype')
```

In addition to extracting genes with a high likelihood, _scVelo_ can provide us with genes whose velocities are different in different cell types. 
Such genes may help to explain the vector field and the inferred lineages. 
The module `tl.rank_velocity_genes` runs a differential velocity t-test and outpus a gene ranking for each cluster.

```{python rank-velocity-genes, class.source = "pythonchunk"}
scv.tl.rank_velocity_genes(adata, groupby = 'celltype', min_corr = 0.3)

df = scv.DataFrame(adata.uns['rank_velocity_genes']['names'])
df.head()

kwargs = dict(frameon = False, size = 20, linewidth = 1.5)

for cluster in ['DIplotene/Secondary spermatocytes', 'Mid Round spermatids']:
  scv.pl.scatter(adata, df[cluster][:5], ylabel = cluster, **kwargs, color = 'celltype')
```

Moreover, partial gene likelihoods (average likelihood over a subset of the cells) can be computed for a each cluster of cells to enable cluster-specific identification of potential drivers.

```{python dynamical-genes, class.source = "pythonchunk"}
scv.tl.rank_dynamical_genes(adata, groupby='celltype')
df = scv.get_df(adata, 'rank_dynamical_genes/names')
df.head(5)
for cluster in ['DIplotene/Secondary spermatocytes', 'Mid Round spermatids']:
    scv.pl.scatter(adata, df[cluster][:5], ylabel = cluster, frameon = False, color = 'celltype')
```


### Additional useful statistics

Finally, we will illustrate two summary statistics, which can be calculated via the `tl.velocity_confidence` module:

* The length of the velocity vector encodes the speed or rate of differentiation. 
* The 'velocity confidence' provides a measure of the coherence of the velocity vector field, that is, how well the velocity vector for a cell correlates with those of its neighbors. 

```{python confidence, class.source = "pythonchunk"}
scv.tl.velocity_confidence(adata)
keys = 'velocity_length', 'velocity_confidence'
scv.pl.scatter(adata, c = keys, cmap = 'coolwarm', perc = [5, 95], basis = 'tsne')
adata.obs
```

## Save the complete AnnData object

Once all calculations are done, we save the `AnnData` object to an .h5ad file. 
This can be read back and further manipulated in python, or imported into R, e.g. using _Seurat_, for further analysis.

```{python save-object, class.source = "pythonchunk"}
adata.write(velodir/'AdultMouseRep3_alevin_GRCm38.gencode.vM21.spliced.intron.fl90.gentrome.k31_sce_nometa_with_velocity.h5ad')
```


# Session info

```{python, class.source = "pythonchunk", eval = TRUE}
from sinfo import sinfo
sinfo()
```

```{r, class.source = "rchunk", eval = TRUE}
sessionInfo()
```

# Other resources

* [_scVelo_ documentation and tutorials](https://scvelo.readthedocs.io/index.html)
* [_velocyto_ documentation](http://velocyto.org/)
* [_alevin_ documentation](https://salmon.readthedocs.io/en/latest/alevin.html)
* [_STARsolo_ documentation](https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md)
* [_kallisto|bustools_ documentation](https://www.kallistobus.tools/)


# References