<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Mike Smith" />


<title>Using HDF5 efficiently</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">On-disk-data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="01-on-disk.html">Introducing HDF5 files</a>
</li>
<li>
  <a href="02-on-disk.html">Exploring single-cell HDF5 file types</a>
</li>
<li>
  <a href="03-on-disk.html">Using HDF5 efficiently</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Using HDF5 efficiently</h1>
<h4 class="author">Mike Smith</h4>

</div>

<div id="TOC">
<ul>
<li><a href="#reading-subsets-of-the-data">Reading subsets of the data</a></li>
<li><a href="#exploring-the-effect-of-chunk-layout">Exploring the effect of chunk layout</a></li>
</ul>
</div>

<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(rhdf5)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw">library</span>(HDF5Array)</span></code></pre></div>
<div id="reading-subsets-of-the-data" class="section level2">
<h2>Reading subsets of the data</h2>
<p>So far we’ve used <code>h5read()</code> or <code>h5dump()</code> to pull the entire contents of a HDF5 dataset in our R session. However this doesn’t take advantage of one of the major features of HDF5 files - efficient access to subsets of a dataset.</p>
<p>What happens if you try to run the code below, which reads the <code>counts_matrix</code> dataset from <code>brain100k.h5</code>? Try using <code>h5ls()</code> to explore the size of the data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>brain_data &lt;-<span class="st"> </span><span class="kw">h5read</span>(<span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data//brain100k.h5&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;/counts_matrix&quot;</span>)</span></code></pre></div>
<p>Sometimes data are simply too large to read into memory. This 100,000 cell matrix is actually a subset of a larger, 1.3 million cell 10X dataset. Naively reading the complete set of counts into a dense R matrix requires ~ 150GB of RAM.</p>
<p>We can use the <code>index</code> argument to specify the elements we want to extract from our dataset. The syntax is a little strange for R, and should be a list with the same length as the number of dimensions in our dataset - in our case that’s two. Each element of the list is a vector providing the indices you want to read or <code>NULL</code> to read everything in that dimension. In the example below we will read all the rows and the first five columns.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>brain_data_subset &lt;-<span class="st"> </span><span class="kw">h5read</span>(<span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data//brain100k.h5&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;/counts_matrix&quot;</span>, <span class="dt">index =</span> <span class="kw">list</span>(<span class="ot">NULL</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>))</span></code></pre></div>
<div id="exercise" class="section level3">
<h3><em>Exercise</em></h3>
<p>Can you modify the code to read other sets of columns? Instead of reading the first five columns try reading the last five or columns 50,001 - 50,005. You can also experiment reading a larger number of columns - perhaps 100, 1,000 or 10,000. Use <code>system.time()</code> to examine how long reading these subsets takes.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co">## Insert your own code here</span></span></code></pre></div>
<p>Hopefully you found that it takes a very similar amount of time to read five consecutive columns from anywhere in the file. You should also notice that the time taken to read larger numbers of columns scales fairly linearly with the number of columns you want.</p>
<p>In this case where our HDF5 file just contains a counts matrix, the HDF5Array approach works equally well.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>HDF5Array<span class="op">::</span><span class="kw">HDF5Array</span>(<span class="dt">filepath =</span> <span class="st">&quot;../data/on-disk-data//brain100k.h5&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;/counts_matrix&quot;</span>)</span></code></pre></div>
<pre><code>## &lt;27998 x 100000&gt; matrix of class HDF5Matrix and type &quot;integer&quot;:
##               [,1]      [,2]      [,3]      [,4] ...  [,99997]  [,99998]  [,99999] [,100000]
##     [1,]         0         0         0         0   .         0         0         0         0
##     [2,]         0         0         0         0   .         0         0         0         0
##     [3,]         0         0         0         0   .         0         0         0         0
##     [4,]         0         0         0         0   .         0         0         0         0
##     [5,]         0         0         0         0   .         0         0         0         0
##      ...         .         .         .         .   .         .         .         .         .
## [27994,]         0         0         0         0   .         0         0         0         0
## [27995,]         1         0         0         2   .         0         1         0         3
## [27996,]         0         0         0         0   .         0         0         0         0
## [27997,]         0         0         0         0   .         0         0         0         0
## [27998,]         0         0         0         0   .         0         0         0         0</code></pre>
</div>
</div>
<div id="exploring-the-effect-of-chunk-layout" class="section level2">
<h2>Exploring the effect of chunk layout</h2>
<p>We’ve established that the chunked nature of HDF5 datasets provides efficient access to subsets of that data. The story is actually more nuanced than this, as HDF5 allows a user to control the size and shape of the chunks when the file is created, which can have a great impact on the performance of any later read operations.</p>
<div class="figure">
<img src="images/Chunk_layout_1.png" alt="In HDF5 datasets can be stored as &quot;chunks&quot; on disk, and if only a subset of data is required only the necessary chunks need to be read.  Chunk layout doesn't have to be symetrical in every dimension, in the 2-dimensional dataset above chunks can consist of entire rows, entire columns, or any other regular partitioning." width="100%" /><img src="images/Chunk_layout_2.png" alt="In HDF5 datasets can be stored as &quot;chunks&quot; on disk, and if only a subset of data is required only the necessary chunks need to be read.  Chunk layout doesn't have to be symetrical in every dimension, in the 2-dimensional dataset above chunks can consist of entire rows, entire columns, or any other regular partitioning." width="100%" />
<p class="caption">
In HDF5 datasets can be stored as “chunks” on disk, and if only a subset of data is required only the necessary chunks need to be read. Chunk layout doesn’t have to be symetrical in every dimension, in the 2-dimensional dataset above chunks can consist of entire rows, entire columns, or any other regular partitioning.
</p>
</div>
<p>We’re going to write the same data multiple times, and take a look at how this effects the read performance. First, lets get hold of a 10,000 cell matrix.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>brain_10k &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">HDF5Array</span>(<span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data/brain100k.h5&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;/counts_matrix&quot;</span>)[,<span class="dv">1</span><span class="op">:</span><span class="dv">10000</span>])</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw">dim</span>(brain_10k)</span></code></pre></div>
<pre><code>## [1] 27998 10000</code></pre>
<p>rhdf5 doesn’t provide access to the chunk dimension settings in <code>h5write()</code>. Instead we have to use a two-step process, creating an empty dataset with <code>h5createDatatset()</code> and then writing to that with <code>h5write()</code>. Here we create the dataset to consist of one single, very large, chunk.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">h5createFile</span>(<span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data/one_chunk.h5&quot;</span>)</span></code></pre></div>
<pre><code>## file &#39;/home/msmith/Teaching/adv_scrnaseq_2020/data/adv_scrnaseq_2020/data/on-disk-data/one_chunk.h5&#39; already exists.</code></pre>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">h5createDataset</span>(<span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data/one_chunk.h5&quot;</span>, </span>
<span id="cb12-2"><a href="#cb12-2"></a>                <span class="dt">dataset =</span> <span class="st">&quot;one_chunk&quot;</span>, </span>
<span id="cb12-3"><a href="#cb12-3"></a>                <span class="dt">dims =</span> <span class="kw">dim</span>(brain_10k), </span>
<span id="cb12-4"><a href="#cb12-4"></a>                <span class="dt">storage.mode =</span> <span class="st">&quot;integer&quot;</span>,</span>
<span id="cb12-5"><a href="#cb12-5"></a>                <span class="dt">chunk =</span> <span class="kw">c</span>(<span class="kw">nrow</span>(brain_10k), <span class="kw">ncol</span>(brain_10k)))</span></code></pre></div>
<pre><code>## You created a large dataset with compression and chunking.
## The chunk size is equal to the dataset dimensions.
## If you want to read subsets of the dataset, you should testsmaller chunk sizes to improve read times.</code></pre>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">h5write</span>(brain_10k, <span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data/one_chunk.h5&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;one_chunk&quot;</span>)</span></code></pre></div>
<div id="exercise-1" class="section level3">
<h3>Exercise</h3>
<p>Try creating other datasets with different chunk sizes. You can pick whatever you like, but I suggest trying square chunks as well as the extremes of single-row or single-column chunks. You can then again use <code>system.time()</code> to see how long it takes to read five columns:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="kw">system.time</span>( <span class="kw">h5read</span>(<span class="dt">file =</span> <span class="st">&quot;../data/on-disk-data/one_chunk.h5&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;/one_chunk&quot;</span>, <span class="dt">index =</span> <span class="kw">list</span>(<span class="ot">NULL</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)) )</span></code></pre></div>
<pre><code>##    user  system elapsed 
##   1.679   1.986   3.602</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
