---
title: "Introducing HDF5 files"
author: "Mike Smith"
output: 
    html_document:
        toc: true
        toc_depth: 2
        toc_float: false
        number_sections: false
        theme: flatly
        highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# HDF5

```{r, message = FALSE}
library(rhdf5)
library(HDF5Array)
```

```{r, fig-1, echo = FALSE, fig.cap = "Example HDF5 file structure"}
knitr::include_graphics('images/hdf5_structure.jpg')
```

## Exploring an HDF5 file

To start with we're going to look at and HDF5 file produced by 10X Genomics (the original file is available from [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k)).  For the moment we aren't going to worry about the specifics of the 10X format, but use this file to demonstrate how you can take a look at the contents of any HDF5 file.

First, lets start with the function `h5ls()`:

```{r, h5ls}
h5ls(file = '../data/pbmc8k_raw_gene_bc_matrices_h5.h5')
```

The output from `h5ls()` gives us an overview of the structure of the file, without really showing us the content.  We can see `pbmc8k_raw_gene_bc_matrices_h5.h5` contains a single group (GRCh38) and within that group there are many datasets.  We can also see what type of data each dataset contains (the `dclass` column).

### Exercise

Use `h5ls()` to examine the other HDF5 files in the `data` folder.

## Reading from HDF5 files


```{r, h5dump}

pbmc8k_list <- h5dump("../data/new_brain.h5")

```


```{r, pbmc1}
tenXdata <- h5read(file = '../data/pbmc8k_raw_gene_bc_matrices_h5.h5', name = "/GRCh38/data")
```









