---
title: "Introducing HDF5 files"
author: "Mike Smith"
output: 
    html_document:
        toc: true
        toc_depth: 2
        toc_float: false
        number_sections: false
        theme: flatly
        highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# HDF5

```{r, message = FALSE}
library(rhdf5)
library(HDF5Array)
```

```{r, fig-1, echo = FALSE, fig.cap = "Example HDF5 file structure"}
knitr::include_graphics('images/hdf5_structure.jpg')
```

## Exploring an HDF5 file

To start with we're going to look at and HDF5 file produced by 10X Genomics (the original file is available from [here](https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k)).  For the moment we aren't going to worry about the specifics of the 10X format, but use this file to demonstrate how you can take a look at the contents of any HDF5 file.

First, lets start with the function `h5ls()`:

```{r, h5ls}
h5ls(file = '../data/pbmc8k_raw_gene_bc_matrices_h5.h5')
```

The output from `h5ls()` gives us an overview of the structure of the file, without really showing us the content.  We can see `pbmc8k_raw_gene_bc_matrices_h5.h5` contains a single group (GRCh38) and within that group there are many datasets.  We can also see what type of data each dataset contains (the `dclass` column).

### Exercise

Use `h5ls()` to examine the other HDF5 files in the `data` folder.

## Reading from HDF5 files


```{r, h5dump}
pbmc8k_list <- h5dump("../data/new_brain.h5")
```

This list is quite large and isn't going to be used again in this tutorial, so I recommend removing it from your R session so you do not run out of RAM.

```{r, cleanup-1}
rm(pbmc8k_list)
```


```{r, pbmc1}
tenXdata <- h5read(file = '../data/pbmc8k_raw_gene_bc_matrices_h5.h5', name = "/GRCh38/data")
```


## Writing to HDF5 files

As you probably expect, it's all possible to write data to HDF5 files.  This is probably a less frequent operation, either because you're processing a large amount of data into something smaller and don't need to use HDF5 to store or another piece of software does hte saving for you, but it useful to know and we will use further examples of this later to explore more properties of HDF5.

In the example below we create a small matrix and write this to a dataset called `example_matrix`.  We then use `h5ls()` to confirm that it's been created.

```{r, write-1}
ex_matrix <- matrix(1:10, nrow = 5, ncol = 2)
h5write(obj = ex_matrix, file = "../data/my_hdf5.h5", name = "example_matrix")
h5ls("../data/my_hdf5.h5")
```


### Exercise 

Try adding other objects to the HDF5 file.  Aspects you can try varying include:
-  R data type e.g. integer, numeric, character, ...
-  R object type e.g. matrix, vector, array, list, data.frame, ...
-  HDF location - you can specify locations in the file hierarchy using `/` e.g. `/place/in/the/file`


## Deleting parts of HDF5 files

Wanting to re-save something is quite a common need - at some point we all realise we've made a mistake or updated some parameters and improved a particular anaylsis. What happens if you try and overwrite an existing group or dataset in an HDF5 file?  

```{r, write-1}
## we've grown out example matrix, and want to resave it.
ex_matrix <- matrix(1:100, nrow = 5, ncol = 20)
h5write(obj = ex_matrix, file = "../data/my_hdf5.h5", name = "example_matrix")
```


```{r h5delete1}
h5ls("../data/my_hdf5.h5", recursive=2)
file.size("../data/my_hdf5.h5")
```


```{r h5delete2}
h5delete(file = "../data/my_hdf5.h5", name = "example_matrix")
h5ls("../data/my_hdf5.h5", recursive=2)
file.size("../data/my_hdf5.h5")
```




