{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Understanding Inferential Replicates\n",
    "\n",
    "We have quantified the [@data source] dataset with alevin and 20 bootstrap inferential replicates.  \n",
    "Let's first understand the basic of bootstrapping, link to slides [@link]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dir <- \"data/pbmc_1k/pbmc_1k\"\n",
    "files <- file.path(dir, \"alevin\", \"quants_mat.gz\")\n",
    "file.exists(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'alevin.log'</li><li>'featureDump.txt'</li><li>'quants_boot_rows.txt'</li><li>'quants_mat_cols.txt'</li><li>'quants_mat.gz'</li><li>'quants_mat_rows.txt'</li><li>'quants_mean_mat.gz'</li><li>'quants_tier_mat.gz'</li><li>'quants_var_mat.gz'</li><li>'whitelist.txt'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'alevin.log'\n",
       "\\item 'featureDump.txt'\n",
       "\\item 'quants\\_boot\\_rows.txt'\n",
       "\\item 'quants\\_mat\\_cols.txt'\n",
       "\\item 'quants\\_mat.gz'\n",
       "\\item 'quants\\_mat\\_rows.txt'\n",
       "\\item 'quants\\_mean\\_mat.gz'\n",
       "\\item 'quants\\_tier\\_mat.gz'\n",
       "\\item 'quants\\_var\\_mat.gz'\n",
       "\\item 'whitelist.txt'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'alevin.log'\n",
       "2. 'featureDump.txt'\n",
       "3. 'quants_boot_rows.txt'\n",
       "4. 'quants_mat_cols.txt'\n",
       "5. 'quants_mat.gz'\n",
       "6. 'quants_mat_rows.txt'\n",
       "7. 'quants_mean_mat.gz'\n",
       "8. 'quants_tier_mat.gz'\n",
       "9. 'quants_var_mat.gz'\n",
       "10. 'whitelist.txt'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"alevin.log\"           \"featureDump.txt\"      \"quants_boot_rows.txt\"\n",
       " [4] \"quants_mat_cols.txt\"  \"quants_mat.gz\"        \"quants_mat_rows.txt\" \n",
       " [7] \"quants_mean_mat.gz\"   \"quants_tier_mat.gz\"   \"quants_var_mat.gz\"   \n",
       "[10] \"whitelist.txt\"       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "system(\"ls data/pbmc_1k/pbmc_1k/alevin\", intern=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have seen the following files before:\n",
    "* `alevin.log`: This file contains the logs generated by alevin while quantifiying the dscRNA-seq data. This is very useful for debugging.\n",
    "* `featureDump.txt`: This is a tsv file with per-cell level summary stats. This is used for whitelisting of the cells.\n",
    "* `quants_mat_cols.txt`: alevin output matrix is cell by gene. This file contains the ordered list of genes denoting the column index in the matrix.\n",
    "* `quants_mat_rows.txt`: alevin output matrix is cell by gene. This file contains the ordered list of cellular barcodes denoting the row index in the matrix.\n",
    "* `quants_mat.gz`: This file contains the output cell-by-gene count matrix in a binary (non human readable) format. We store it in a compressed EDS format which can be efficiently stored and loaded back into memory. We show some some of the benchmark comparing common single-cell format below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this alevin run we have bunch of new files:\n",
    "* `quants_mean_mat.gz`: The files contains the `mean` of the bootstrap estimates and the values are stored in the EDS format just like `quants_mat.gz`.\n",
    "* `quants_var_mat.gz`: The files contains the `variance` of the bootstrap estimates and the values are stored in the EDS format just like `quants_mat.gz`.\n",
    "* `whitelist.txt`: Alevin performs an inteliigent whitelisting procedure using features from the `featureDump.txt` file to classify the quantified barcodes \n",
    "into high or low confidence barcodes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing alevin data with tximeta\n",
    "\n",
    "We will use *tximeta* to import the *alevin* counts into\n",
    "R/Bioconductor. The main function `tximeta` reads information from the\n",
    "entire output directory of *alevin* or *Salmon* in order to\n",
    "automatically detect and download metadata about the reference\n",
    "sequences (the transcripts). It should work \"out of the\n",
    "box\" for human, mouse, and fruit fly reference transcripts from\n",
    "GENCODE, Ensembl, or RefSeq.\n",
    "\n",
    "First we specify the path where the quantification data is stored. In\n",
    "this tutorial, the data is already copied to your environment.\n",
    "\n",
    "**Note:** For typical use, you would just specify the path to the directory \n",
    "of the output from *alevin*.\n",
    "\n",
    "**Note:** when sharing *alevin* or *Salmon* datasets, make sure\n",
    "to share the entire output directory (you can `zip` or `tar` the\n",
    "directories to make them into a single share-able file).\n",
    "\n",
    "We make a vector called `files` (for *alevin* this will point to just\n",
    "a single file, but with *Salmon* usually one would import multiple\n",
    "files at a time)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can import the *alevin* quantification using the following call to\n",
    "`tximeta`. The extra argument to *alevin* will filter out cells based\n",
    "on *alevin*'s post-quantification quality control methods (see paper\n",
    "for details)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "importing quantifications\n",
      "\n",
      "reading in alevin gene-level counts across cells with fishpond\n",
      "\n",
      "filtering down to 1142 cell barcodes\n",
      "\n",
      "reading in alevin inferential variance with fishpond\n",
      "\n",
      "found matching transcriptome:\n",
      "[ GENCODE - Homo sapiens - release 33 ]\n",
      "\n",
      "building TxDb with 'GenomicFeatures' package\n",
      "\n",
      "Import genomic features from the file as a GRanges object ... \n",
      "OK\n",
      "\n",
      "Prepare the 'metadata' data frame ... \n",
      "OK\n",
      "\n",
      "Make the TxDb object ... \n",
      "Warning message in .get_cds_IDX(mcols0$type, mcols0$phase):\n",
      "\"The \"phase\" metadata column contains non-NA values for features of type\n",
      "  stop_codon. This information was ignored.\"\n"
     ]
    }
   ],
   "source": [
    "library(tximeta)\n",
    "se <- tximeta(files, type=\"alevin\", alevinArgs=list(filterBarcodes=TRUE))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see as it was importing that it automatically detected we are\n",
    "working with data quantified using GENCODE's *Homo sapiens* reference\n",
    "transcripts, release 33. The transcript metadata was added\n",
    "programmatically to the output (this object, `se`). `tximeta` is also\n",
    "written in such a way that it will avoid unnecessary downloads and\n",
    "parsing of files -- it will first attempt to find a cached version of\n",
    "the metadata, to avoid re-downloading or re-parsing these files.\n",
    "\n",
    "`tximeta` returns a *SummarizedExperiment*. We can\n",
    "easily convert this object into a *SingleCellExperiment*\n",
    "which has specific slots designed for single-cell\n",
    "experiment data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library(SingleCellExperiment))\n",
    "sce <- as(se, \"SingleCellExperiment\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The SingleCellExperiment object is used widely across Bioconductor packages (as you may already know), and so the code up to this point can be used as an entry point into other Bioconductor single cell workflows. For more details on working with SingleCellExperiment objects, one can consult the following online book: Orchestrating Single-Cell Analysis with Bioconductor."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add cell annotations\n",
    "\n",
    "To visualize the inferential replicates, let's first cluster and extract the expected cell-types for the cellular barcodes. We use Seurat to cluster and marker genes to assign cell-types.\n",
    "We start by first extracting the count matrix from the _SingleCellExperiment_ object and generate the _Seurat_ object from that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages(library(dplyr))\n",
    "library(Seurat)\n",
    "\n",
    "# create seurat object for whitelisted cells\n",
    "cts <- assays(sce)[[\"counts\"]]\n",
    "pbmc <- CreateSeuratObject(cts)\n",
    "pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We perform some basic filtering to remove noisy cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mt.genes <- rownames(sce)[as.logical(seqnames(sce) == \"chrM\")]\n",
    "pbmc[[\"percent.mt\"]] <- PercentageFeatureSet(pbmc, features = mt.genes)\n",
    "VlnPlot(pbmc, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3)\n",
    "\n",
    "# filtering some outlier cells\n",
    "pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 5000 & percent.mt < 20 & nCount_RNA < 20000)\n",
    "pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We normalize the data, extract PCA embeddings, run clustering algorithm & generate UMAP embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize and runPCA\n",
    "pbmc <- FindVariableFeatures(pbmc) %>%  NormalizeData() %>%  ScaleData(verbose = F) %>% RunPCA(verbose = F)\n",
    "\n",
    "# Cluster the cells\n",
    "pbmc <- FindNeighbors(pbmc, dims = 1:25, verbose = F)\n",
    "pbmc <- FindClusters(pbmc, resolution = 0.15, verbose = F)\n",
    "\n",
    "# plotting umap embeddings\n",
    "pbmc <- RunUMAP(pbmc, dims = 1:25, verbose = F)\n",
    "DimPlot(pbmc, label = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We find the marker genes using the standard Seurat workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find marker genes & extract top 10 genes\n",
    "pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)\n",
    "top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)\n",
    "DoHeatmap(pbmc, features = top10$gene) + NoLegend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the marker genes we assign each cluster their expected cell-types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assigning Cluster names based on marker genes\n",
    "# \"CD14+ Monocytes\" \"CST3\" \"ENSG00000101439.9\"\n",
    "# \"CD4 T\" \"IL7R\" \"ENSG00000168685.15\"\n",
    "# \"B\" \"MS4A1\" \"ENSG00000156738.17\"\n",
    "# \"CD8 T\" \"KLRB1\" \"ENSG00000111796.4\" (we use only for this particular set of cells)\n",
    "# \"NK\" \"GNLY\" \"ENSG00000115523.16\"\n",
    "\n",
    "markers.vec <- c(\"CD14+ Monocytes\"=\"ENSG00000101439.9\", \"CD4 T\"=\"ENSG00000168685.15\", \"B\"=\"ENSG00000156738.17\", \"CD8 T\"=\"ENSG00000111796.4\", \"NK\"=\"ENSG00000115523.16\")\n",
    "clusters <- character(5)\n",
    "for (i in seq_along(markers.vec)) {\n",
    "  idx <- which(top10$gene == markers.vec[i])\n",
    "  stopifnot(length(idx) == 1)\n",
    "  cluster.idx <- as.numeric(as.character(top10$cluster[idx])) + 1\n",
    "  clusters[cluster.idx] <- names(markers.vec)[i]\n",
    "}\n",
    "\n",
    "top10$cluster.name <- rep(clusters, each=10)\n",
    "df <- as.data.frame(top10)\n",
    "# write.csv(df, file=\"../extdata/top10.csv\")\n",
    "\n",
    "names(clusters) <- levels(pbmc)\n",
    "pbmc <- RenameIdents(pbmc, clusters)\n",
    "# saveRDS(Idents(pbmc), file=\"../extdata/idents.rds\")\n",
    "\n",
    "DimPlot(pbmc, reduction = \"umap\", label = TRUE, pt.size = 0.5) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since not all cell had been We subset to the cells that we have IDs for, and attach the labels in\n",
    "the correct order:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ids <- Idents(pbmc)\n",
    "\n",
    "idx <- colnames(sce) %in% names(ids)\n",
    "table(idx)\n",
    "sce <- sce[,idx]\n",
    "sce$cluster <- ids[colnames(sce)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of cells per cluster:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table(sce$cluster)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting counts with uncertainty\n",
    "\n",
    "In this section, we will begin to plot the counts for cells, for\n",
    "specific genes, and showing the *inferential uncertainty* as\n",
    "quantified by *alevin*. This is a unique aspect to the *alevin*\n",
    "quantification method, that it retains gene multi-mapping reads,\n",
    "instead of discarding these reads. *alevin* can also attach a measure\n",
    "of uncertainty to each count in the matrix. *alevin* computes the mean\n",
    "and variance of *inferential replicates* which are generated by\n",
    "bootstrapping the read data. `tximeta` will import these inferential\n",
    "mean and variance matrices by default (they are also sparse, similar\n",
    "to the counts matrix).\n",
    "\n",
    "We will first visualize the uncertainty, and later give an\n",
    "example of a set of genes where the uncertainty is indicating\n",
    "significant quantification uncertainty which is useful to consider\n",
    "when performing EDA (exploratory data analysis) or statistical\n",
    "analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o <- order(colSums(assays(sce)[[\"counts\"]]), decreasing=TRUE)\n",
    "sce <- sce[,o]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now use a plotting function `plotInfReps` from the *fishpond*\n",
    "package in order to plot the inferential mean and variance for the\n",
    "cells, for various genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(fishpond)\n",
    "plotInfReps(sce, idx=\"ENSG00000167286.9\", x=\"cluster\",\n",
    "            legend=TRUE, reorder = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we have x as a grouping variable for coloring the points, one can also specify x to be a continuous covariate, e.g. pseudo-time. An additional covariate cov can be used for additional grouping, e.g. batches, donors or lineages.\n",
    "\n",
    "Here we show the same plot but now subsetting the number of cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx <- sample(ncol(sce),200)\n",
    "plotInfReps(sce[,idx], idx=\"ENSG00000167286.9\",\n",
    "            x=\"cluster\", legend=TRUE, reorder = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This time only 100 cells:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx <- sample(ncol(sce),100)\n",
    "plotInfReps(sce[,idx],\n",
    "            idx=\"ENSG00000167286.9\",\n",
    "            x=\"cluster\",\n",
    "            legend=TRUE, reorder = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inferential variance (uncertainty)\n",
    "\n",
    "In this section, we will briefly talk about how the\n",
    "inferential uncertainty as stored in the variance assay may be useful\n",
    "in practice.\n",
    "\n",
    "Many users may prefer to just work with the counts matrix, and not\n",
    "consider the inferential mean and variance. We have found that,\n",
    "globally, this may not lead to too much of a problem, but for certain\n",
    "genes, it may be important to use the inferential variance in cases\n",
    "where it may signal difficult to quantify genes. Inferential\n",
    "uncertainty for bulk and single cell RNA-seq differential expression\n",
    "was a focus of the *Swish* nonparametric statistical method [@swish]\n",
    "which we do not demonstrate here, but one can refer to\n",
    "the vignette of [swish](https://bioconductor.org/packages/swish) for\n",
    "more details (in particular the section on *alevin* data).\n",
    "\n",
    "Let's start by visualizing the uncertainty across all values in the\n",
    "matrix: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "var <- as.vector(assays(sce)[[\"variance\"]])\n",
    "mu <- as.vector(assays(sce)[[\"mean\"]])\n",
    "idx <- mu > 3\n",
    "df <- data.frame(log10mean=log10(mu[idx]),\n",
    "                 log10var=log10(var[idx]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "ggplot(df, aes(log10mean, log10var)) +\n",
    "  geom_hex(bins=100) + \n",
    "  geom_abline(intercept=0, slope=1, col=\"red\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The values around the red line indicate not much uncertainty (bootstrapping reads gives us roughly multinomial, and so also approximately Poisson, variability around the mean count). However we can see a tail of higher uncertainty values in the matrix, where the inferential variance is for example, up to 10 times higher than the mean.\n",
    "\n",
    "We can also plot this summarized to a single value per gene, here we calculate the 99% quantile of the ratio of variance over mean, per gene, and plot this over the mean:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(matrixStats)\n",
    "rratio <- rowQuantiles(as.matrix(\n",
    "  assays(sce)[[\"variance\"]] /\n",
    "  (assays(sce)[[\"mean\"]] + 1)), probs=.99)\n",
    "rmu <- rowMeans(assays(sce)[[\"mean\"]])\n",
    "idx <- rmu > .1\n",
    "df <- data.frame(log10mean=log10(rmu[idx]),\n",
    "                 rratio=rratio[idx],\n",
    "                 gene=mcols(sce)$gene_id[idx])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A number of genes have a ratio above 2, but then two genes in this dataset stand out above the rest, and they happen to be a ribosomal gene, and another gene that is a read-through of the same locus:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with(df, plot(log10mean, rratio))\n",
    "high.uncert <- which(df$rratio > 10)\n",
    "with(df[high.uncert,],\n",
    "     points(log10mean, rratio, pch=20, col=\"red\"))\n",
    "with(df[high.uncert,],\n",
    "     text(log10mean, rratio, gene, pos=c(2,4)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcols(sce)$gene_id[idx][high.uncert]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of the counts for the ribosomal gene and its read-through:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "par(mfrow=c(2,1), mar=c(2.5,4.5,1,1))\n",
    "plotInfReps(sce[,1:1000], idx=\"ENSG00000265681.7\",\n",
    "            x=\"cluster\", reorder=T)\n",
    "plotInfReps(sce[,1:1000], idx=\"ENSG00000215472.10\",\n",
    "            x=\"cluster\", reorder=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The extra uncertainty on the counts for this gene indicate that alevin was not certain if the reads should come from the gene or its read-through. While in this case, the ribosomal gene may not be over interest, there are other cases (e.g. developmental or immune genes with high sequence homology) where information of the uncertainty of quantification can be useful in interpreting the data. For example, the uncertainty among the marker genes is in reasonable bounds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "par(mfrow=c(3,2), mar=c(2.5,4.5,1,1))\n",
    "plotInfReps(sce[,1:1000], idx=markers.vec[1],\n",
    "            x=\"cluster\", reorder=T)\n",
    "plotInfReps(sce[,1:1000], idx=markers.vec[2],\n",
    "            x=\"cluster\", reorder=T)\n",
    "plotInfReps(sce[,1:1000], idx=markers.vec[3],\n",
    "            x=\"cluster\", reorder=T)\n",
    "plotInfReps(sce[,1:1000], idx=markers.vec[4],\n",
    "            x=\"cluster\", reorder=T)\n",
    "plotInfReps(sce[,1:1000], idx=markers.vec[5],\n",
    "            x=\"cluster\", reorder=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
